pkg/controller/migplan/migplan_controller.go

```diff
@@ -145,7 +145,27 @@ func (r *ReconcileMigPlan) Reconcile(request reconcile.Request) (reconcile.Resul
 		}
 	}
 
-	if plan.Status.HasCriticalCondition() {
+	// plan is closed, ensure resource cleanup
+	if plan.Spec.Closed {
```

Comment:
Seems like the `closed` behavior should happen before everything else in the reconcile().  Even before the validation.  If the plan is closed, I don't think we want to reconcile anything else.  also, even if `ensureClosed()` succeeds, we should return.

Reply:
We don't return on success because immediately following this block, the next block calls SetReady(false) on the plan and returns. If we return here, we'll have to add another SetReady(false) call -- which will work fine, but might be harder to follow. If we decide to add the SetReady here, along with the extra return, we could call it before validate. I guess there's no harm, for closed plans, to skip validating src/dest clusters, etc., especially since we're no longer explicitly referencing src/dest clusters in the delete anymore, since we delete from all clusters.

Reply:
I just noticed that you added the additional `plan.Spec.Closed` to the `plan.Status.HasCriticalCondition()` below.  This address my primary concern (and motivation for suggesting the return) that the reconcile would continue into the methods for ensuring the repository and storage.  So not returning will be fine.

_snip_
```
if plan.Spec.Closed || plan.Status.HasCriticalCondition() {
```
As for validating a closed plan - I guess it depends on how may closed plans will exist at any given time.  I'd imagined 40-50 (closed) plans all doing needless validation at potentially high rates given the local and remote watches will still be wired up.  Perhaps I'm overestimating the performance impact or the number of closed plans but this seems fairly undesirable.  Another issue with running this after the validations is that the condition _staging_ used by `validate()` will always clear the `Closed` condition.  If need, I have thoughts on how this can be dealt with but should probably be done in a separate PR.

I agree that if it's behavior is moved to be beginning of the reconcile() and the return added, the `Ready` condition will need to be explicitly cleared.

I also noticed that nothing is checking for the `Closed` condition so the resource clean up will be always performed on closed plans.  Even after everything has been cleaned up.

Reply:
So regarding checking the closed condition. Yeah, right now this keeps checking (and presumably not finding) things to delete. We should probably change the `if plan.Spec.Closed` Condition to `if plan.Spec.Closed && !plan.Status.HasCondition(Closed)` -- that would only run this if the plan has been flagged for closure via `Spec.Closed` but for which closure has not been completed.

Reply:
As for validation on closed plans, I just looked back at the validate function. I was initially concerned that a plan that failed validation shouldn't attempt to close, in case something was missing that validation checked, but at least for registry validation, I don't think that's the case. We're just iterating over all Clusters -- so validate isn't even looking at clusters that have things we need to clean up but aren't connected to the plan anymore. There are 2 ways to skip validation on closed plans:
1) Move this block of code up above validation -- then we will need to add the SetReady(false) and return after ensureClosed -- and we'll have to revisit the complexity around multiple places for errors to pop up -- since we have to check for error on ensureClosed *and* try to call setReady.
2) keep the code as-is, but wrap the validate call in an `if !plan.Spec.Closed` conditional
It's not obvious to me which approach is cleaner.

Reply:
When I recommended putting the _close_ logic in close.go I had envisioned _something_ like this (meant to be pseudo-code):

_reconcile loop_
```

// Closed
closed, err := r.handleClosed(plan)
if err != nil {
    if errors.IsConflict(err) {
        return reconcile.Result{Requeue: true}, nil
    } else {
        return reconcile.Result{}, err
    }
} 
if closed {
    return reconcile.Result{}, nil
}

// Validations.
err = r.validate(plan)
...
```

_close.go_
```
func (r ReconcileMigPlan) handleClosed(plan *migapi.MigPlan) (bool, error) {
    closed := plan.Spec.Closed
    if !closed || plan.Status.hasCondition(Closed) {
        return closed, nil
    }
  
    plan.Status.SetReady(false)
    err = r.Update(context.TODO(), plan)
    if err != nil {
        return closed, err
    }

    err := r.ensureClosed(plan)
    return closed, err
}
```

Complete delegation to another method provides for simpler logi and also leaves all of the code in `Reconcile()` after `handleClosed()` concerned with an _open_ plan.  Continuing into the `validate()` will clear the `Closed` condition in _re-open_ scenarios.

Reply:
"Continuing into the validate() will clear the Closed condition in re-open scenarios."
Is that because validate always clears all conditions and starts over? There's nothing in validate() that touches the Closed condition explicitly. In any case, this will be easy enough to verify. If the above approach handles reopening, then setting Spec.Closed to false will trigger re-opening and recreating the registry.

Reply:
Yes.  

The condition _staging_ which includes BeginStagingConditions() and BeginStagingConditions() used in validation will result in the delete any conditions that have not been explicitly: SetCondition().  This approach was indented to simplify managing conditions in validation.  Since then, other conditions have been added that are not managed as part of validation.  The side effect is that non-validation conditions get cleared too.  

I will likely open a PR today to add support for (optionally) limiting _staging_ to specific condition categories with the default being those categories managed by validation.  Namely: Critical, Error, Warn.  This will make the behavior explicit, as well as, configurable.  The `Closed` condition is currently category=Critical which we _could_ reconsider changing but fine the way it is.